### DISCUSSION TITLE
How to pass state down a long chain of functions?

### DISCUSSION THREAD

**Discussion by schoening** [pelme/htpy#46] [General]
Having to pass state down from the `body -> side_menu -> current_tab -> list_of_folders -> folder -> list_of_items -> item` isn't fun.
How do you solve this?

**Comment by pelme** [comment]
React has useContext etc for this(https://react.dev/learn/passing-data-deeply-with-context). I have been thinking of doing something similar for htpy but I think using the builtin contextvars module (https://docs.python.org/3/library/contextvars.html) solves this without doing anything else in htpy:

```py

import contextvars

import htpy as h

color_scheme = contextvars.ContextVar("color_scheme", default="light")


print(h.div(class_=f"color-scheme-{color_scheme.get()}")["Hi"])

color_scheme.set("dark")

print(h.div(class_=f"color-scheme-{color_scheme.get()}")["Hi"])
```

result:

```
<div class="color-scheme-light">Hi</div>
<div class="color-scheme-dark">Hi</div>
```
So it is basically global state but managed a bit nicer. By using contextvars, it would be thread/async safe and works with typing. set/reset can be used in a view/middleware to set the values.

Does that work for you? In that case we should definitely document this pattern.

**Comment by pelme** [comment]
(Did not yet had the need for this myself so I have not used it.)

**Comment by pelme** [comment]
I played with making it possible to pass contexts similar to how React does it, what do you think about #48 ?

**Comment by schoening** [comment]
Ohh! Thats very nice of you 😃

Could you write up an example how this would be used in a scenario where I have multiple nested components?

# Example:
I have a theme (str) and a side_menu_open (bool).
How do I access those two using the react style use_context "hook" 😄?

```
@app.get("/")
def get_root(request: Request):
    return index_page()
```

```
def index_page():
    return (
        body(...)[
            side_menu(),
            main
    )

def side_menu():
    return aside(...)
```

**Reply by pelme** [reply]
The idea is that you first define your context variable such as

```py
from htpy import Context
theme_ctx = Context("gold")
```

You provide/override values for all sub-components with `theme_ctx.provide("silver", lambda: my_component())`. (lambda/callable must be used).

To get/use a value, use `theme_ctx.use(lambda value: div["the current theme value is {value}"`


Your example would look something like this (I also added type hints but those are optional):

```py
import htpy as h

# Defaults are currently required but should be optional (and raise an error if trying to use them without providing a value).
theme_ctx = h.Context("plain")
side_menu_open_ctx = h.Context(False)


# @app.get("/")
def get_root() -> h.Node:
    current_theme = "blue"  # get user theme from database/session etc
    side_menu_open = True  # get the side menu open state from database/session etc

    return theme_ctx.provide(
        current_theme,
        lambda: side_menu_open_ctx.provide(side_menu_open, lambda: index_page()),
    )


def index_page() -> h.Node:
    return theme_ctx.use(
        lambda theme: h.body(class_=f"theme-{theme}")[
            side_menu(),
            h.main(),
        ]
    )


def side_menu() -> h.Node:
    return side_menu_open_ctx.use(lambda is_open: h.aside(open=is_open)["The side menu"])


print(get_root())
```

output:
```
$ python context_example.py
<body class="theme-blue"><aside open>The side menu</aside><main></main></body>
```

**Reply by schoening** [reply]
Cool!
Is that less work than using the build in python context directly?

Have you seen ReactPy? They have a hooks api. I haven't looked at that library yet I only skimmed the docs.

**Reply by pelme** [reply]
contextvars works too but is more like using globals. Using it properly could be tricky and have a lot of rough edges with unexpected behavior when using lazy callables/generators and overriding the context in part of the tree. It can be useful for instance to set the request object in a middleware and then be able to access it everywhere. It is just a global variable that is thread/asyncio safe. So if that is all you need, contextvars is a fine solution.

Here is an example where using a generator(lazy evaluated) and trying to override the theme for a part of the tree:

```py
import contextvars

import htpy as h

theme = contextvars.ContextVar("theme", default="plain")


# @app.get("/")
def get_root() -> h.Node:
    return index_page()


def index_page() -> h.Node:
    token = theme.set("fancy")
    result = h.body(class_=f"theme-{theme.get()}")[
        h.h1[f"Showing of the {theme.get()} theme!"],
        h.ul[(item() for x in range(2))],
    ]
    theme.reset(token)
    return result


def item() -> h.Node:
    return h.li[f"Theme is currently {theme.get()}"]


print(get_root())

```

which outputs:
```
<body class="theme-fancy"><h1>Showing of the fancy theme</h1><ul><li>Theme is currently plain</li><li>Theme is currently plain</li></ul></body>
```
(notice that the value is reverted to "plain" in the `<li>`s when the rendering happens.)

There are many ways to shoot oneself in the foot like that. That is why I am playing with a htpy-rendering-tree-aware-Context-API. That kind of bug/confusion can never happen then. The context values then lives entirely in the htpy rendering tree and always matches it.

Thanks for showing ReactPy, I had not seen it before! Creating hooks would require adding a decorator to the component in any case I think. Also, there will probably never be any other hooks in htpy(useState, useEffect etc does not make sense here). Anyways, it got me thinking and we could have a decorator that adds the context value as an argument instead:

```py
@side_menu_open_ctx.consumer
def side_menu(side_menu_is_open: bool) -> h.Node:
    return h.aside(open=side_menu_is_open)["The side menu"]
```

I think it looks quite a bit nicer, what do you think about that?

**Comment by schoening** [comment]
Thanks for the explanation!

I agree that you probably don't want hooks, etc. in htpy.
People can use ReactPy if they want that 😄.

Love the decorator yes 👏🏻!

**Comment by pelme** [comment]
This is released as part of htpy 24.8.2, give it a try and let me know how it goes!
https://htpy.dev/usage/#passing-data-with-context

**Reply by schoening** [reply]
Super awesome that you provided a feature this quickly!

Question:

How do I pass multiple different contexts to a page?

The side_menu is going to need both the `theme_ctx` and also the `tree_ctx`;
How should I pass the data?

```
def side_menu(theme: Theme, tree: FileAndFolderTree) -> h.Node:
    return ...
```

**Reply by pelme** [reply]
There are two options:

- If different context values always are provided and consumed together, consider creating a class/dataclass to wrap them both to avoid creating "too many" contexts.
- You can consume multiple contexts in the same component by using the decorator multiple times. The decorator will add the value as the *first argument*. Regular Python rules about decorator ordering applies so the decorator with the first argument goes last in the list. In your case:

```py
@tree_context.consumer
@theme_context.consumer
def side_menu(theme: Theme, tree: FileAndFolderTree) -> h.Node:
    return ...
```

**Reply by pelme** [reply]
(Documented the fact that a component can have multiple consumers: https://htpy.dev/usage/#passing-data-with-context)

**Reply by schoening** [reply]
I have been testing it out all day!
It's a good feature, otherwise I wouldn't be coding on a job project at 19:30 😆

~~But I am noticing it gets very slow when I try to pass several contexts.~~

~~I have not yet investigated if that is due to the amount of data I am pulling from the db or the iterations.~~

**Reply by pelme** [reply]
That is nice to hear! Have not had any really good use case for them myself yet :)

I did a quick run of the `scripts/benchmark_big_table.py` and could not see any difference from before contexts. That script does not use the contexts though. Let me know if you find something that suggests contexts would slow down the rendering, I don't think they should add much overhead though!

**Reply by schoening** [reply]
Looks like the perf issues where due to my db code and network.
This is a great addition to the library, I appreciate it!

I am surprised you do not have a use-case for it in your own code.
Do you not need to pass state down a long hierarchy of components in your codebase?
Do you have any code examples that you are allowed to share?

**Comment by schoening** [comment]
Just FYI, we ended up not using this after all 😅

**Reply by pelme** [reply]
thanks for the update! how did you solve the problem instead?

**Reply by schoening** [reply]
We ended up building a class based view. With render functions and the contexts living directly on the object.
